### PR TITLE
[WIP] Check monitor for checkpoints every epoch even if there is no validation

### LINKED ISSUE #4603: ModelCheckpoint misbehaves when no validation
## 🐛 Bug

ModelCheckpoint misbehaves when there is no validation step. Eg. if you ask to monitor for loss and save every checkpoint with `save_top_k=-1` the callback will not do this silently.

### To Reproduce
This [line of code](https://github.com/PyTorchLightning/pytorch-lightning/blob/470e2945fc0fefa1b6d42f1d64fef9f22f5c75b3/pytorch_lightning/callbacks/model_checkpoint.py#L180) is a pure indicator of failure without a validation step.

### Expected behavior

Callback being called after every epoch

### Environment

latest pytorch-lightning

### Additional context
I hoped to finetune the model a bit and aggregate the last checkpoints to do SWA. I've fixed the problem using the modified version of `ModelCheckpoint`

```python
from pytorch_lightning.callbacks.model_checkpoint import ModelCheckpoint as _ModelCheckpoint


class ModelCheckpoint(_ModelCheckpoint):
    def on_epoch_end(self, trainer, pl_module):
        super().on_validation_end(trainer, pl_module)

    def on_validation_end(self, trainer, pl_module):
        pass
```

I can open a PR if there is no concern about this kind of change

### PR DESCRIPTION
**Pull request by ferrine**
## What does this PR do?

Check monitor for checkpoints every epoch even if there is no validation

Fixes #4603 

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In in short, see following bullet-list:  

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified; _Bugfixes should be including in bug-fix release milestones (m.f.X) and features should be included in (m.X.b) releases._
 

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by Borda** [comment]
@ferrine it seems it breaks a few tests, but first we shall agree on this API change...
cc: @PyTorchLightning/core-contributors

**Comment by rohitgr7** [comment]
doesn't this line of code calls the checkpoint if there is no validation
https://github.com/PyTorchLightning/pytorch-lightning/blob/94a9d3d2837eb962cb47ad2854569039a552f729/pytorch_lightning/trainer/training_loop.py#L614-L615

if not can you create a reproducible notebook using [bug_report_model](https://colab.research.google.com/drive/1HvWVVTK8j2Nj52qU4Q4YCyzOm0_aLQF3?usp=sharing)

**Review by awaelchli — CHANGES_REQUESTED** [review]
this change is incorrect. the tests are failing for good reason :) 
As @rohitgr7 said, let's find out what the actual problem is in your use case.

**Comment by ferrine** [comment]
@awaelchli, @rohitgr7 Thanks for the review, I think I have an idea where the error happens. I'll revert the commit and add my failing use case as the next step

**Comment by ferrine** [comment]
@rohitgr7 I've created a [notebook](https://colab.research.google.com/drive/1oc_19GlqoNp2hRPfEWapjFUuOebV5EiU?usp=sharing) reproducing the error

TL;DR Error reproduces if the following conditions hold
* You override validation_step
* You do not provide val_dataloader

**Comment by rohitgr7** [comment]
> @rohitgr7 I've created a [notebook](https://colab.research.google.com/drive/1oc_19GlqoNp2hRPfEWapjFUuOebV5EiU?usp=sharing) reproducing the error
> 
> TL;DR Error reproduces if the following conditions hold
> 
>     * You override validation_step
> 
>     * You do not provide val_dataloader

@ferrine 
That's why it doesn't hit this condition.
https://github.com/PyTorchLightning/pytorch-lightning/pull/4793#issuecomment-731341084

I think the main issue is here:
https://github.com/PyTorchLightning/pytorch-lightning/blob/8601268c70649f49767001098adbf665a93843df/pytorch_lightning/trainer/training_loop.py#L565-L569

the checkpoint callback is supposed to be called after each validation generally. This can be called multiple times within an epoch or at the epoch end. The problem here is that the call that is supposed to be made at the end of the epoch, with validation, happens here only so https://github.com/PyTorchLightning/pytorch-lightning/pull/4793#issuecomment-731341084 is added to do a checkpoint when there is no validation at the epoch end.

Another way we can handle this is by imposing a condition that if a `validation_step` is overridden then a `val_dataloader` should be there with an exception otherwise. But this is not convenient so there is just a warning for that.

One solution I suggest a better way to handle this is by allowing checkpoint required within an epoch to happen here, otherwise the checkpoint that is supposed to happen at the end of the epoch should happen outside this loop with no conditions related to overridden stuff at all. For this we need to refactor the below method in two different methods., one for step related and another one for epoch related. This will solve a few underlying issues I am aware of right now.
https://github.com/PyTorchLightning/pytorch-lightning/blob/8601268c70649f49767001098adbf665a93843df/pytorch_lightning/trainer/training_loop.py#L842-L851

Not sure if it will break anything.

**Comment by awaelchli** [comment]
Could one do the following: Run the checkpoint on every step (train, val, test regardless). Internally in ModelCheckpoint, keep track of last global step and/or last epoch when we saved a checkpoint to avoid checkpointing multiple times on the same step. Then this logic would be independent of training loop and entirely live inside ModelCheckpoint. Whether or not a checkpoint gets saved will be determined by a) is monitor key available in logged metrics b) have we saved on the current global step already c) ... maybe more conditions

**Comment by rohitgr7** [comment]
> a) is monitor key available in logged metrics

@awaelchli 
if we implement this logic and someones doesn't log the metric specified in the monitor, it should give a proper warning or an exception. How will you distinguish this condition in these two different case?

**Comment by tchaton** [comment]
Any update on this PR ?

**Comment by Borda** [comment]
@ferrine seems that the PR is empty... is it intended or some other problems with GH?

**Comment by ferrine** [comment]
There are no problems with the PR itself, I usually plan to code for weekends. Traditional lack of time because of 5/2 dayjob:))

**Comment by ferrine** [comment]
> Whether or not a checkpoint gets saved will be determined by a) is monitor key available in logged metrics b) have we saved on the current global step already c) ... maybe more conditions?

Seems like b)  this is already in place:
https://github.com/PyTorchLightning/pytorch-lightning/blob/9b1afa8c877e36851f93903f99aa6951953b9148/pytorch_lightning/callbacks/model_checkpoint.py#L228

**Comment by ferrine** [comment]
I observe `trainer.global_step` increment between the last `on_epoch_end` and `on_train_end`, is that normal? This results in duplicated checkpoints

**Comment by ferrine** [comment]
Seems like I solved the increment problem in the same wy it was before. Now I have Another error in

https://github.com/PyTorchLightning/pytorch-lightning/blob/9b1afa8c877e36851f93903f99aa6951953b9148/tests/checkpointing/test_checkpoint_callback_frequency.py#L97

with `{'epochs': 2, 'val_check_interval': 0.3, 'expected': 7}`

As far as I understand, this results in this checkpoint schedule:
```

0.3: checkpoint
0.3: checkpoint
0.3: checkpoint

0.3: checkpoint
0.3: checkpoint
```

**Comment by ferrine** [comment]
Hey, this seems to be ready for review. I also noticed some inconsistency in `rank_zero` usage: the callback is supposed to be used with rank zero only, but a lot of methods are evaluated on other ranks. Any suggestions on what I should do with them?

**Comment by awaelchli** [comment]
> but a lot of methods are evaluated on other ranks. Any suggestions on what I should do with them?

yes this is correct and needs to stay this way. 
the state of the object needs to be kept in sync, even if rank > 0 does not save anything to disk.

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 1]
```diff
@@ -503,11 +530,17 @@ def _add_backward_monitor_support(self, trainer):
         if self.save_top_k is None and self.monitor is not None:
             self.save_top_k = 1
 
+    def _valid_monitor_key(self, trainer):
```
Comment (carmocca) [inline pytorch_lightning/callbacks/model_checkpoint.py:533]: Maybe merge together this function and `_is_valid_monitor_key`?

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 2]
```diff
@@ -545,6 +578,7 @@ def _save_last_checkpoint(self, trainer, pl_module, ckpt_name_metrics):
 
         # when user ALSO asked for the 'last.ckpt' change the name
         if self.save_last:
+            rank_zero_info("Saving latest checkpoint...")
```
Comment (carmocca) [inline pytorch_lightning/callbacks/model_checkpoint.py:581]: This message will appear every time the checkpoint logic is run.

We only want it to appear the last time (as it did before)

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 3]
```diff
@@ -215,22 +229,35 @@ def on_load_checkpoint(self, checkpointed_state: Dict[str, Any]):
         self.best_model_score = checkpointed_state["best_model_score"]
         self.best_model_path = checkpointed_state["best_model_path"]
 
-    def save_checkpoint(self, trainer, pl_module):
+    def should_save(self, trainer, is_last=False):
+        epoch = trainer.current_epoch
+        global_step = trainer.global_step
+        should_save = not (
+            # negative conditions
+            self.save_top_k == 0 or
```
Comment (carmocca) [inline pytorch_lightning/callbacks/model_checkpoint.py:237]: About formatting: `black` is not strictly enforced but generally used in this project. The issue here is that you put your logical operators at the end whereas black puts them at the beginning. This means that this function will look quite different after somebody comes around and does automatic formatting.

Can you do it yourself? `black -S -l 120 model_checkpoint.py` and then fix the comment positions and other undesired changes in formatting

Comment (ferrine) [reply pytorch_lightning/callbacks/model_checkpoint.py:237]: Yeah, black was used before, but CI formatter had a different opinion. I'll figure it out

Comment (carmocca) [reply pytorch_lightning/callbacks/model_checkpoint.py:237]: The CI formatter probably complained about this: https://www.flake8rules.com/rules/W503.html

But the rule is outdated as mentioned in the link

**Comment by ferrine** [comment]
I'll be back from winter holidays in a week 🙂

**Comment by tchaton** [comment]
Hey @ferrine,

Any updates ?

Best,
T.C

**Comment by ferrine** [comment]
Yes, I'm back from mountains just yesterday. I'll be back to the PR on
weekends 🙂

File: pytorch_lightning/callbacks/model_checkpoint.py (hunk 2 above)
Comment (ferrine) [reply pytorch_lightning/callbacks/model_checkpoint.py]: I've added a condition to call this info only in `on_training_end`

**Comment by ferrine** [comment]
I've merged master into this branch and observe a bunch of horovod related errors in some checks, is that ok?

**Comment by carmocca** [comment]
> I've merged master into this branch and observe a bunch of horovod related errors in some checks, is that ok?

Yes, we have some issues with Horovod, mostly using torch==1.3. They should be fixed in the feature branch.

This makes me think, is this PR more of a bugfix or a feature? It seems to me like it is a feature in which case it should point to the `release/1.2-dev` branch. cc: @Borda

File: pytorch_lightning/trainer/training_loop.py [hunk 4]
```diff
@@ -606,9 +587,6 @@ def run_training_epoch(self):
             self.num_optimizers
         )
 
-        # when no val loop is present or fast-dev-run still need to call checkpoints
-        self.check_checkpoint_callback(not (should_check_val or is_overridden('validation_step', model)))
```
Comment (rohitgr7) [inline pytorch_lightning/trainer/training_loop.py:610]: you can just fix your usecase by adding `sum(self.trainer.num_val_batches) == 0` here. working on a PR https://github.com/PyTorchLightning/pytorch-lightning/pull/5208 fixing more issues there.

**Comment by rohitgr7** [comment]
@carmocca it's a bug since this is a case we missed for no validation checkpoint.

**Comment by ferrine** [comment]
Anything I can help with?

File: pytorch_lightning/trainer/training_loop.py (hunk 4 above)
Comment (ferrine) [reply pytorch_lightning/trainer/training_loop.py:610]: @rohitgr7, does it make sense to close the issue in #5208, not here? I'm fine with closing the PR if a nicer solution is proposed

Comment (rohitgr7) [reply pytorch_lightning/trainer/training_loop.py:610]: I'd suggest yes since doing a bit of a refactor there to fix more issues. Your use-case is already fixed there. Mind check if it works for you??
